### PR TITLE
Generate EPUB and MOBI ebooks

### PR DESCRIPTION
**Pull request by sapegin**
As discussed in #273.

It’s far from perfect (for example, no covers) but very readable on both iPad and Kindle.

ebooks

So now the question is how to generate / update the files. I see two possible solutions:

1. Generate files on CI / build (will make build slower).
2. Commit ebook files to the repo (they are big).

What do you think?

### CODE REVIEW

**Comment by m1guelpf** [comment]
@sapegin Maybe we can push it to releases instead of pushing them to the repo...

**Comment by bkeepers** [comment]
🎉  This is awesome! Thanks for taking a crack at this.

> So now the question is how to generate / update the files.

…

> Maybe we can push it to releases instead of pushing them to the repo...

☝️ this is what I was thinking. We could start doing regular releases where we generate the book in various formats.

File: script/ebooks [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+const download = require('download');
+const ebrew = require('ebrew');
+
+const processor = unified()
+  .use(parse)
+  .use(replaceAsides)
+  .use(replaceCheckboxes)
+  .use(fixImages)
+  .use(downloadImages)
+  .use(stringify);
```
Comment (bkeepers) [inline script/ebooks]: Have you experimented with ways of converting the HTML to epub? I wonder if that would give better output and be easier to maintain long term.

I tried [node-html2epub](https://github.com/fabi1cazenave/node-html2epub), which has a really [simple way of declaring the contents](https://github.com/fabi1cazenave/node-html2epub/blob/master/examples/epub3.json) and will fetch content from the web, but the ebook it generates is corrupt and the project doesn't appear to be maintained.

File: script/ebooks [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+function downloadImages(text) {
+  function image(processor) {
+    return ast => visit(ast, 'image', node => {
+      if (node.url.startsWith('http')) {
+        const url = node.url;
+        const filename = slug(path.basename(node.url));
+        const filepath = `./.asset-downloads/${filename}.jpg`
+        node.url = filepath;
+        if (fs.existsSync(filepath)) {
+          promises.push(Promise.resolve());
```
Comment (wooorm) [inline script/ebooks]: @sapegin Why is it needed to add a resolving promise here?

Would `if (!fs.existsSync(filepath)) { download() ...` not be simpler?

File: script/ebooks [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  '_articles/starting-a-project.md',
+  '_articles/finding-users.md',
+  '_articles/building-community.md',
+  '_articles/best-practices.md',
+  '_articles/leadership-and-governance.md',
+  '_articles/getting-paid.md',
+  '_articles/code-of-conduct.md',
+  '_articles/metrics.md',
+  '_articles/legal.md',
+]);
```
Comment (wooorm) [inline script/ebooks]: Could we read the `_articles` directory here, so we don’t need to update this list?

File: script/ebooks [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+const visit = require('unist-util-visit');
+const download = require('download');
+const ebrew = require('ebrew');
+
+const processor = unified()
+  .use(parse)
+  .use(replaceAsides)
+  .use(replaceCheckboxes)
+  .use(fixImages)
+  .use(downloadImages)
```
Comment (wooorm) [inline script/ebooks]: Should `downloadImages` be used as a plugin? It looks like it has a different signature: https://github.com/github/open-source-guide/pull/305/files#diff-a7359ca92a3e17674bcedbb64dae862dR109

**Comment by wooorm** [comment]
Awesome stuff, @sapegin!

**Comment by wooorm** [comment]
> > Maybe we can push it to releases instead of pushing them to the repo...
>
> ☝️ this is what I was thinking. We could start doing regular releases where we generate the book in various formats.

Great idea! I think this can even be automated, with [Travis x GitHub Releases](https://docs.travis-ci.com/user/deployment/releases/)!

File: script/ebooks (hunk 1 above)
Comment (sapegin) [inline script/ebooks]: In that case we’ll need to generate a separate set of HTML files for the book and make a custom CSS to style asides, etc. Tooling for generating ebooks are far from great unfortunately.

File: script/ebooks (hunk 4 above)
Comment (sapegin) [inline script/ebooks]: Nope, that’s a bug, thanks! It’s user in `buildBooks`.

File: script/ebooks (hunk 3 above)
Comment (sapegin) [inline script/ebooks]: Then you’ll need a way to order documents…

File: script/ebooks (hunk 2 above)
Comment (sapegin) [inline script/ebooks]: Yeah, good point!

**Comment by sapegin** [comment]
> Great idea! I think this can even be automated, with Travis x GitHub Releases!

Sounds good! Probably should be done not in this PR. It also requires API keys.

File: script/ebooks (hunk 3 above)
Comment (wooorm) [inline script/ebooks]: Ah, right! 👍

File: script/ebooks (hunk 1 above)
Comment (bkeepers) [inline script/ebooks]: Eventually I think I'd like to go that route, but we can do something simpler for now to get it working and evolve it later.

Comment (sapegin) [inline script/ebooks]: Yeah, in long term it has benefits like better control on the style and formatting, custom CSS, etc. But for an MVP Markdown is enough I think ;-)

**Comment by bkeepers** [comment]
This is looking pretty good! I opened #316 to track improvements that we want to make, but am good with this as v1.

**Comment by Johnstone456** [comment]
#2652

**Comment by Johnstone456** [comment]
Dbgg

**Pull request merged**